src/object/obj_tx.c

```diff
@@ -640,154 +673,774 @@ dc_tx_open(tse_task_t *task)
+}
+
+static int
+dc_tx_commit_prepare(struct dc_tx *tx, tse_task_t *task)
+{
+	daos_unit_oid_t			 leader_oid = { 0 };
+	struct daos_csummer		*csummer;
+	struct dc_tx_req_group		*dtrgs = NULL;
+	struct daos_cpd_sub_head	*dcsh = NULL;
+	struct daos_cpd_disp_ent	*dcdes = NULL;
+	struct daos_shard_tgt		*shard_tgts = NULL;
+	struct dtx_memberships		*mbs;
+	struct dtx_daos_target		*ddt;
+	struct dc_tx_rdg		*dtr;
+	void				*ptr;
+	d_list_t			 dtr_list;
+	daos_size_t			 sgl_size = 0;
+	size_t				 size;
+	uint32_t			 leader_dtrg_idx = 0;
+	uint32_t			 act_tgt_cnt = 0;
+	uint32_t			 act_grp_cnt = 0;
+	uint32_t			 start;
+	uint32_t			 tgt_cnt;
+	uint32_t			 req_cnt;
+	int				 rc = 0;
+	int				 i;
+	int				 j;
+
+	D_INIT_LIST_HEAD(&dtr_list);
+	csummer = dc_cont_hdl2csummer(tx->tx_coh);
+	req_cnt = tx->tx_read_cnt + tx->tx_write_cnt;
+	tgt_cnt = pool_map_target_nr(tx->tx_pool->dp_map);
+	D_ASSERT(tgt_cnt != 0);
+
+	if (tx->tx_flags & DAOS_TF_RDONLY)
+		start = tx->tx_total_slots - tx->tx_read_cnt;
+	else if (tx->tx_total_slots > DTX_SUB_WRITE_MAX)
+		start = tx->tx_total_slots - DTX_SUB_WRITE_MAX - tx->tx_read_cnt;
+	else
+		start = (tx->tx_total_slots >> 1) - tx->tx_read_cnt;
+
+	D_ALLOC_ARRAY(dtrgs, tgt_cnt);
+	if (dtrgs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	for (i = 0; i < req_cnt; i++) {
+		struct daos_cpd_sub_req	*dcsr = &tx->tx_req_cache[i + start];
+		struct dc_object	*obj = dcsr->dcsr_obj;
+
+		if (dcsr->dcsr_opc == DCSO_UPDATE) {
+			rc = dc_tx_classify_update(tx, dcsr, csummer);
+			if (rc < 0)
+				goto out;
+
+			sgl_size += rc;
+			if (sgl_size >= OBJ_BULK_LIMIT) {
+				rc = dx_tx_bulk_prepare(dcsr, task);
+				if (rc != 0)
+					goto out;
+			} else {
+				dcsr->dcsr_update.dcu_sgls = dcsr->dcsr_sgls;
+			}
+		}
+
+		if (dcsr->dcsr_opc == DCSO_PUNCH_OBJ) {
+			for (j = 0; j < obj->cob_grp_nr; j++) {
+				rc = dc_tx_classify_common(tx, dcsr, dtrgs,
+						tgt_cnt, j, i, false, true,
+						&leader_dtrg_idx, &act_tgt_cnt,
+						&dtr_list, &leader_oid);
+				if (rc != 0)
+					goto out;
+			}
+		} else {
+			rc = obj_dkey2grpidx(obj, dcsr->dcsr_dkey_hash,
+					     tx->tx_pm_ver);
+			if (rc < 0)
+				goto out;
+
+			rc = dc_tx_classify_common(tx, dcsr, dtrgs, tgt_cnt, rc,
+					i, dcsr->dcsr_opc == DCSO_READ, false,
+					&leader_dtrg_idx, &act_tgt_cnt,
+					&dtr_list, &leader_oid);
+			if (rc != 0)
+				goto out;
+		}
+	}
+
+	size = dc_tx_reduce_rdgs(&dtr_list, &act_grp_cnt);
+	size += sizeof(*ddt) * act_tgt_cnt;
+
+	D_ALLOC_PTR(dcsh);
+	if (dcsh == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC(dcsh->dcsh_mbs, size + sizeof(*dcsh->dcsh_mbs));
+	if (dcsh->dcsh_mbs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(dcdes, act_tgt_cnt);
+	if (dcdes == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	D_ALLOC_ARRAY(shard_tgts, act_tgt_cnt);
+	if (shard_tgts == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	dcsh->dcsh_xid = tx->tx_id;
+	dcsh->dcsh_leader_oid = leader_oid;
+	dcsh->dcsh_epoch = tx->tx_epoch.oe_value;
```

Comment:
Is dcsh supposed to contain the epoch context for one TX? If so, we will also need to pack tx_epoch.oe_first and tx_epoch.oe_flags (after our current PRs have landed, of course).

Reply:
Yes, once your patch, the "dcsh_epoch" will be redefined from "daos_epoch_t" to "struct dtx_epoch".

Reply:
Sounds good.